### PR TITLE
Report diagnostics from `TryFindDisposePatternMethod` when binding `foreach` loop

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/30257
Fixes https://github.com/dotnet/roslyn/issues/73934

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 1]
```diff
@@ -4860,9 +4863,52 @@ public sealed class AsyncEnumerator : System.IAsyncDisposable
                 Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.MoveNextAsync()").WithLocation(7, 15),
                 // (7,15): warning CS0612: 'C.AsyncEnumerator.Current' is obsolete
                 //         await foreach (var i in new C())
-                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.Current").WithLocation(7, 15)
-                );
-            // Note: Obsolete on DisposeAsync is not reported since always called through IAsyncDisposable interface
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4865]: So the comment was incorrect - we do not go through IAsyncDisposable interface?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 2]
```diff
@@ -4860,9 +4863,52 @@ public sealed class AsyncEnumerator : System.IAsyncDisposable
...
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.Current").WithLocation(7, 15));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/30257")]
+        public void TestWithPatternAndObsolete_WithoutDisposableInterface()
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4870]: Consider testing with explicit interface implementation of the DisposeAsync method.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4865]: Yes, this was changed in [PR](https://github.com/dotnet/roslyn/pull/72598) but this comment in test was missed

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs [hunk 3]
```diff
@@ -1193,15 +1193,16 @@ private void GetDisposalInfoForEnumerator(SyntaxNode syntax, ref ForEachEnumerat
...
-                MethodSymbol patternDisposeMethod = TryFindDisposePatternMethod(receiver, syntax, isAsync, BindingDiagnosticBag.Discarded, out bool expanded);
+                BindingDiagnosticBag patternDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
+                MethodSymbol patternDisposeMethod = TryFindDisposePatternMethod(receiver, syntax, isAsync, patternDiagnostics, out bool expanded);
                 if (patternDisposeMethod is object)
                 {
                     Debug.Assert(!patternDisposeMethod.IsExtensionMethod);
                     Debug.Assert(patternDisposeMethod.ParameterRefKinds.IsDefaultOrEmpty ||
                         patternDisposeMethod.ParameterRefKinds.All(static refKind => refKind is RefKind.None or RefKind.In or RefKind.RefReadOnlyParameter));
 
+                    diagnostics.AddRangeAndFree(patternDiagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:1205]: >diagnostics.AddRangeAndFree(patternDiagnostics); [](http://example.com/codeflow?start=20&length=48)

I am curious what effect this is going to have on collection expressions and params collections. Is this diagnostics going to be meaningful for them? Consider adding tests.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 4]
```diff
@@ -4860,9 +4863,52 @@ public sealed class AsyncEnumerator : System.IAsyncDisposable
...
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.DisposeAsync()").WithLocation(7, 15),
+                // (7,15): warning CS0612: 'C.GetAsyncEnumerator(CancellationToken)' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.GetAsyncEnumerator(System.Threading.CancellationToken)").WithLocation(7, 15),
+                // (7,15): warning CS0612: 'C.AsyncEnumerator.MoveNextAsync()' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.MoveNextAsync()").WithLocation(7, 15),
+                // (7,15): warning CS0612: 'C.AsyncEnumerator.Current' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.Current").WithLocation(7, 15));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4911]: >Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.Current").WithLocation(7, 15)); [](http://example.com/codeflow?start=16&length=118)

Consider reverting changes on this line and reducing unnecessary diff.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 5]
```diff
@@ -8503,8 +8552,7 @@ public static class Extensions
                 Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.Enumerator.MoveNextAsync()").WithLocation(8, 15),
                 // (8,15): warning CS0612: 'C.Enumerator.Current' is obsolete
                 //         await foreach (var i in new C())
-                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.Enumerator.Current").WithLocation(8, 15)
-                );
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.Enumerator.Current").WithLocation(8, 15));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:8555]: >); [](http://example.com/codeflow?start=127&length=2)

Consider reverting changes on this line and reducing unnecessary diff.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs [hunk 6]
```diff
@@ -9424,10 +9424,12 @@ public static class CExt
...
             comp.VerifyDiagnostics(
-                // (14,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static non-abstract, non-virtual methods or static local functions.
+                // 0.cs(7,9): error CS8901: 'SEnumerator.Dispose()' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
+                //         foreach (var i in s) {}
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "foreach (var i in s) {}").WithArguments("SEnumerator.Dispose()").WithLocation(7, 9),
+                // 0.cs(14,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static non-abstract, non-virtual methods or static local functions.
                 //     [UnmanagedCallersOnly]
-                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, "UnmanagedCallersOnly").WithLocation(14, 6)
-            );
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, "UnmanagedCallersOnly").WithLocation(14, 6));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:9432]: >); [](http://example.com/codeflow?start=120&length=2)

Consider reverting changes on this line and reducing unnecessary diff.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs [hunk 7]
```diff
@@ -9424,10 +9424,12 @@ public static class CExt
 ", UnmanagedCallersOnlyAttribute });
 
             comp.VerifyDiagnostics(
-                // (14,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static non-abstract, non-virtual methods or static local functions.
+                // 0.cs(7,9): error CS8901: 'SEnumerator.Dispose()' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
+                //         foreach (var i in s) {}
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "foreach (var i in s) {}").WithArguments("SEnumerator.Dispose()").WithLocation(7, 9),
+                // 0.cs(14,6): error CS8896: 'UnmanagedCallersOnly' can only be applied to ordinary static non-abstract, non-virtual methods or static local functions.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:9430]: >0.cs [](http://example.com/codeflow?start=19&length=4)

Consider reverting changes on this line and reducing unnecessary diff.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs (hunk 2 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4870]: Did you add this test? I couldn't find it.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4870]: My mistake. Added now :-)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs [hunk 8]
```diff
@@ -9424,10 +9424,44 @@ public static class CExt
...
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/73934")]
+        public void UnmanagedCallersOnlyDeclaredOnPatternDispose_CollectionExpression()
+        {
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+public struct S
+{
+    public static void M2(S s)
+    {
+        int[] a = [42, ..s];
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:9445]: >..s [](http://example.com/codeflow?start=23&length=3)

This is an interesting scenario, but it is not quite the scenario I had in mind. I was referring to a check for a creatable collection, which involves determination of an element type (foreach element type).

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs [hunk 9]
```diff
@@ -4852,6 +4852,54 @@ public sealed class AsyncEnumerator : System.IAsyncDisposable
...
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.DisposeAsync()").WithLocation(7, 15),
+                // (7,15): warning CS0612: 'C.GetAsyncEnumerator(CancellationToken)' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.GetAsyncEnumerator(System.Threading.CancellationToken)").WithLocation(7, 15),
+                // (7,15): warning CS0612: 'C.AsyncEnumerator.MoveNextAsync()' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.MoveNextAsync()").WithLocation(7, 15),
+                // (7,15): warning CS0612: 'C.AsyncEnumerator.Current' is obsolete
+                //         await foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbol, "foreach").WithArguments("C.AsyncEnumerator.Current").WithLocation(7, 15));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs:4866]: >); [](http://example.com/codeflow?start=132&length=2)

Consider reverting changes on this line and reducing unnecessary diff.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:9445]: I didn't follow. How could that involve binding the `Dispose` method?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:9445]: > I didn't follow. How could that involve binding the `Dispose` method?

It looks like ```TryFindDisposePatternMethod``` might be reachable from ```TryGetCollectionIterationType```.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:9445]: Added a test. The behavior seems fine (because `TryGetCollectionIterationType` discards diagnostics)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8)
